sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,285 @@
+    for available_key in public_key_list.lines() {
+        let f: Vec<&str> = available_key.split('=').collect();
+        // If f[0] isn't a number, then it isn't a valid index.
+        if f[0].parse::<u32>().is_ok() {
+            let public_key_endpoint = format!("{}/{}/openssh-key", IMDS_PUBLIC_KEY_BASE_URL, f[0]);
+            public_key_endpoints.push(public_key_endpoint);
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid index. Skipping...",
+                &f[0]
+            );
+            continue;
+        }
+    }
+    if public_key_endpoints.is_empty() {
+        warn!("No valid key endpoints found");
+    }
+    public_key_endpoints
+}
+
+/// Store the args we receive on the command line.
+struct Args {
+    log_level: LevelFilter,
+}
+
+/// Print a usage message in the event a bad arg is passed
+fn usage() {
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!(
+        r"Usage: {}
+            [ --log-level trace|debug|info|warn|error ]",
+        program_name
+    );
+}
+
+/// Parse the args to the program and return an Args struct
+fn parse_args(args: env::Args) -> Result<Args> {
+    let mut log_level = None;
+
+    let mut iter = args.skip(1);
+    while let Some(arg) = iter.next() {
+        match arg.as_ref() {
+            "--log-level" => {
+                let log_level_str = iter.next().ok_or_else(|| error::Error::Usage {
+                    message: "Did not give argument to --log-level".into(),
+                })?;
+                log_level = Some(LevelFilter::from_str(&log_level_str).map_err(|_| {
+                    error::Error::Usage {
+                        message: format!("Invalid log level '{}'", log_level_str),
+                    }
+                })?);
+            }
+
+            _ => {
+                return Err(error::Error::Usage {
+                    message: "unexpected argument".into(),
+                })
+            }
+        }
+    }
+
+    Ok(Args {
+        log_level: log_level.unwrap_or_else(|| LevelFilter::Info),
+    })
+}
+
+fn run() -> Result<()> {
+    let args = parse_args(env::args())?;
+
+    // TerminalMode::Stderr will send all logs to stderr, as sundog only expects the json output of
+    // the setting on stdout.
+    TermLogger::init(args.log_level, LogConfig::default(), TerminalMode::Stderr)
+        .context(error::Logger)?;
+
+    info!("shibaken started");
+
+    let public_keys = fetch_public_keys_from_imds()?;
+
+    let user_data = UserData::new(public_keys);
+
+    info!("Generating user-data");
+    // Serialize user_data to a JSON string that can be read by the admin container.
+    let user_data_json = serde_json::to_string(&user_data).context(error::SerializeJson)?;
+    debug!("{}", &user_data_json);
+
+    info!("Encoding user-data");
+    // Encode user_data as base64 to represent it in the base64 `user-data` setting, which allows
```

Comment:
nit: This comment could probably be reduced to something like `Host container user-data must be base64 encoded`

Reply:
Agree with @zmrow here - this is a constraint from the model, not a choice of the generator.